### ISSUE TITLE
Maximum visible window limit (option)

### ISSUE THREAD

**Issue by MutatingFunc** [ianyh/Amethyst#1075]
**Is your feature request related to a problem? Please describe.**
Minor issue - windows inevitably become unusably small if you have a ton of them open, meaning you have to manually clear them away.

**Describe the solution you'd like**
Personally, I'd choose to have the oldest windows be automatically minimised. This would also work more reliably, and be less risky, than closing them completely, which may trigger save prompts.

The user would choose how many windows they're comfortable with, accounting for their personal preference & monitor size.

**Describe alternatives you've considered**
If it's not too difficult, the user could choose between having the oldest windows be minimised or closed.

**Additional context**
This would enable an iPad-like workflow on macOS, which is kind of nice for monotaskers. Relatedly, this would work well as a focus tool. Floating windows would not count toward the limit, making them analogous to iPad slideover windows.

If anyone else likes this idea, and wants to try this out, check out my Proof Of Concept branch here: https://github.com/MutatingFunc/Amethyst/tree/POC-autominimize
I'll be using this branch & keeping this branch roughly in sync with develop for the forseeable future until [the PR](https://github.com/ianyh/Amethyst/pull/1083) is merged.

**Comment by macintacos** [comment]
I only just found out about Amethyst, and this was pretty much the very first thing that I wanted. I guess the related PR is stalled?

**Comment by MutatingFunc** [comment]
Yeah, it was 90% there, I just needed some pointers as to best practice for this project. I'd still be happy to finish this off.

I suppose after a year it couldn't hurt to have another go - @ianyh is this feature of interest to you?

**Comment by Bellavene** [comment]
May I ask you the compiled version? Tried to compile it on M1, with no success. Something with cocoapods.
Thank you

**Comment by MutatingFunc** [comment]
I've heard your request. I'll send an update when I've had time to focus on this again and resurrect the branch. I've just been a little swamped in the last few weeks!

**Comment by MutatingFunc** [comment]
For those interested in trying this branch, I've uploaded a build on my fork of the repo. Try it out here:
https://github.com/MutatingFunc/Amethyst/releases/tag/v1.XX.1

**Comment by Bellavene** [comment]
Thank you

**Comment by Bellavene** [comment]
Thank you! Works great. Would be awesome if you could add an option to deminimize a window if you close one. At this point I have setup a hotkey to do this in hammerspoon:

```
hs.hotkey.bind({ 'cmd', 'shift' }, 'm', function()
  for _, win in ipairs(hs.window.allWindows()) do
    if win:isMinimized() then
      win:unminimize()
      win:focus()
      break
    end
  end
end)
```

Finder's desktop is treated as a window and in that case, if you close or minimize Finders last window focus stays on Finder app, very annoying. There is a workaround for it, also in hammerspoon:

```
wf = hs.window.filter
allwindows = wf.new(nil)
allwindows:subscribe({wf.windowDestroyed, wf.windowMinimized, wf.windowHidden}, function(window, appName, reason)
    local app2 = hs.application.frontmostApplication()
    local bundle = app2:bundleID()
    local count = 0

    for k, v in pairs(app2:visibleWindows()) do
            count = count + 1
    end

    if app2:bundleID() == "com.apple.finder" then
        if count < 2 then
            app2:activate(true)
            app2:hide()
            return
        end
    end
end)
```

**Comment by MutatingFunc** [comment]
I agree, un-minimising the last window would be awesome. There's this cool macOS window concept, "Flows", which sort of follows this philosophy of working with a "flow" of app windows (in spirit at least): https://uxdesign.cc/my-attempt-at-redesigning-the-desktop-experience-macos-case-study-99f5f2fb3b10

However, since I don't have a clue what I'm doing with the Amethyst codebase **_and_** I'm really stretching what the project is designed to do, right now I'd just be happy just getting the basic window limit in without turning the codebase into spaghetti. I got really lucky that there was an unused "minimise window" helper function just lying around, and I don't think Amethyst currently tracks minimised windows so I'd be messing (more than I am already) with some pretty fundamental stuff. 😅

**Comment by Bellavene** [comment]
Totally understandable :) Anyway, it is great even as is now

**Comment by Bellavene** [comment]
> I agree, un-minimising the last window would be awesome. There's this cool macOS window concept, "Flows", which sort of follows this philosophy of working with a "flow" of app windows (in spirit at least): https://uxdesign.cc/my-attempt-at-redesigning-the-desktop-experience-macos-case-study-99f5f2fb3b10

Thats near the same as PaperWM. There is a script for hammerspoon. https://github.com/mogenson/PaperWM.spoon
But sadly after playing with it, I found it too buggy to be my main driver, Amethyst works better than anything out there for now.

**Comment by ianyh** [comment]
I think this has been implemented by #1319
